core/coreapi/coreapi.go

```diff
@@ -15,26 +15,70 @@ package coreapi
 
 import (
 	"context"
+	"errors"
+	"fmt"
 
-	core "github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core"
 	coreiface "github.com/ipfs/go-ipfs/core/coreapi/interface"
-
+	"github.com/ipfs/go-ipfs/core/coreapi/interface/options"
+	"github.com/ipfs/go-ipfs/namesys"
+	"github.com/ipfs/go-ipfs/pin"
+	"github.com/ipfs/go-ipfs/repo"
+
+	ci "gx/ipfs/QmNiJiXwWE3kRhZrC5ej3kSjWHm337pYfhjLGSCDNKJP2s/go-libp2p-crypto"
+	exchange "gx/ipfs/QmP2g3VxmC7g7fyRJDj1VJ72KHZbJ9UW24YjSWEj1XTb4H/go-ipfs-exchange-interface"
+	pstore "gx/ipfs/QmQAGG1zxfePqj2t7bLxyN8AFccZ889DDR9Gn8kVLDrGZo/go-libp2p-peerstore"
+	blockstore "gx/ipfs/QmS2aqUZLJp8kF1ihE5rvDGE5LvmKDPnx32w9Z1BW9xLV5/go-ipfs-blockstore"
+	record "gx/ipfs/QmSoeYGNm8v8jAF49hX7UwHwkXjoeobSrn9sya5NPPsxXP/go-libp2p-record"
+	bserv "gx/ipfs/QmVDTbzzTwnuBwNbJdhW3u7LoBQp46bezm9yp4z1RoEepM/go-blockservice"
+	offlinexch "gx/ipfs/QmYZwey1thDTynSrvd6qQkX24UpTka6TFhQ2v569UpoqxD/go-ipfs-exchange-offline"
+	routing "gx/ipfs/QmZBH87CAPFHcc7cYmBqeSQ98zQ3SX9KUxiYgzPmLWNVKz/go-libp2p-routing"
+	p2phost "gx/ipfs/QmahxMNoNuSsgQefo9rkpcfRFmQrMN6Q99aztKXf63K7YJ/go-libp2p-host"
+	pubsub "gx/ipfs/Qmc3BYVGtLs8y3p4uVpARWyo3Xk2oCBFF1AhYUVMPWgwUK/go-libp2p-pubsub"
 	ipld "gx/ipfs/QmcKKBwfz6FyQdHR2jsXrrF6XeSBXYL86anmWNewpFpoF5/go-ipld-format"
+	peer "gx/ipfs/QmcqU6QUDSXprb1518vYDGczrTJTyGwLG9eUa5iNX4xUtS/go-libp2p-peer"
 	logging "gx/ipfs/QmcuXC5cxs79ro2cUuHs4HQ2bkDLJUYokwL8aivcX6HW3C/go-log"
 	dag "gx/ipfs/QmdURv6Sbob8TVW2tFFve9vcEWrSUgwPqeqnXyvYhLrkyd/go-merkledag"
+	offlineroute "gx/ipfs/QmdxhyAwBrnmJFsYPK6tyHh4Yy3gK8gbULErX1dRnpUMqu/go-ipfs-routing/offline"
 )
 
 var log = logging.Logger("core/coreapi")
 
 type CoreAPI struct {
-	node *core.IpfsNode
-	dag  ipld.DAGService
+	nctx context.Context
+
+	identity   peer.ID
+	privateKey ci.PrivKey
+
+	repo       repo.Repo
+	blockstore blockstore.GCBlockstore
+	baseBlocks blockstore.Blockstore
+	pinning    pin.Pinner
+
+	blocks bserv.BlockService
+	dag    ipld.DAGService
+
+	peerstore       pstore.Peerstore
+	peerHost        p2phost.Host
+	recordValidator record.Validator
+	exchange        exchange.Interface
+
+	namesys namesys.NameSystem
+	routing func(bool) (routing.IpfsRouting, error)
+
+	pubSub *pubsub.PubSub
+
+	// TODO: this can be generalized to all functions when we implement some
+	// api based security mechanism
+	isPublishAllowed func() error
```

Comment:
This method isn't really about security. Really, we *should* allow publishing IPNS records while /ipns is mounted. We don't currently because we haven't implemented that yet.

Reply:
I know, I just wanted to hint that something like this may be useful (but a full security layer implementing coreapi interface would probably be better)